### ISSUE TITLE
Performance issue with newer versions of Faunus

### ISSUE THREAD

**Issue by mlund** [mlund/faunus#214]
Below is a benchmark of the `examples/bulk` system of melted NaCl for the same number particles, trial moves, release mode; and all analysis disabled. Hardcoded, non-splined potentials are used in both cases. 

1. faunus-v1.0.2:  **54 secs**
2. faunus-master (f5e11ef10107d11a441b2c1dd08906cf860fc2bd): **105 secs**

Possible places to look:
- [x] `geometry::vdist()`
- [ ] passing of `change`
- [x] compare with nonbonded switched off, i.e. probe only trial move code
- [ ] compare clang and gcc

ping @rc83

**Comment by mlund** [comment]
Letting `Potential::Coulomb` and `WeeksChandlerAndersen` return constant 1.0, only reduces the time to 80 secs, indicating that the bottleneck is not the pair-potential. Completely skipping `g_internal()` in Nonbonded, the simulation finishes in a few seconds. Next target, `vdist()`.

**Comment by rc83** [comment]
My guess based on some quick tests: The difference between 75 s and 105 s is at least partly caused by (not) inlining of the `vdist` function. When I accidentally kept some garbage in a branch of `vdist` which was not executed at all, the run time increased from 75 s to 90 s in my case.

**Comment by mlund** [comment]
On my system, omitting the first `ORTHOGONAL` if-stement reduces time quite to ~75 secs which could be the threshold for the compiler to use inlining. Still, I'm surprised that the the raw `a-b` distance is so much quicker, i.e. 35 secx. I'd like to try with GCC, but I currently cannot make it link on the mac setup (brew or cleanup may be needed, hrmf).

**Comment by rc83** [comment]
Clang inlines the `Chameleon::vdist` function when the following patch is applied. I have a three seconds difference between hardcoded conditions for orthogonal coordinates and PBC and their runtime evaluation. However there is still a 15 seconds penalty for evaluating the PBC.
 
```diff --git a/src/geometry.h b/src/geometry.h
--- a/src/geometry.h
+++ b/src/geometry.h
@@ -397,10 +397,6 @@ inline Point Chameleon::vdist(const Point &a, const Point &b) const {
             else if (distance.z() < -len_half.z())
                 distance.z() += len.z();
         }
-    } else if (boundary_conditions.coordinates == ORTHOHEXAGONAL ||
-               boundary_conditions.coordinates == TRUNC_OCTAHEDRAL) {
-        distance = a - b;
-        boundary(distance);
     } else {
         distance = geometry->vdist(a, b);
     }
```

**Comment by mlund** [comment]
How do you check if it's inlined using clang? I suppose code size affect the decision for inlining but it should be possible to allow for a more aggressive setting. Also, I read that LTO could be important, but I have not tested. Yes, PBC seems to imply a significant cost; perhaps we should have a look at the two papers to see if there's a faster approach. A number of other functions in geometry are inlined, but need not be (`getVolume()` etc.)

**Comment by rc83** [comment]
I checked it by indirect observations. I need to set up profiler to do optimizations more seriously. I do not think that our PBC algorithm is slow by itself. My wild guess is that the increase in size of the `vdist` function may prevent another inlining in the chain (imagine a matryoshka doll). I suggest following:

- To use gcc with the `-Winline` warning to clean the code from unused `inline` keywords and find candidates for possible `[[gnu:always_inline]]` attribute.
- To profile the code.

**Comment by mlund** [comment]
Part of this has been addressed in #231.

**Comment by mlund** [comment]
After commit #231 I did a new test of a bulk program, similar to the above (LJ particles) with equal number of particles; moves; no analysis and got the following:

 mode |    v1.0.2 |  master (87daa57)
 ----      | ------ | --------
Release | 25 secs     |  18 secs
RelWithDebInfo | 34 secs |  28 secs
